rfcs/20190502-preprocessing-layers.md

```diff
@@ -0,0 +1,578 @@
+
+**Separate training mechanism**
+
+The internal state of a `PreprocessingLayer` is not affected by backpropagation: all of its weights are non-trainable. A `PreprocessingLayer` has to be trained in a separate step, as follow:
+
+```python
+preprocessing_layer.adapt(data_sample)
+```
+
+**Possible non-differentiability**
+
+Processing layers extend Keras by allowing preprocessing to be part of the model. Unlike existing layers, these computations are not always differentiable, e.g. both `Discretize` and `VectorizeText` are non-differentiable.
+
+As a result, all preprocessing layers are treated as frozen when used as part of a model. In addition, if a non-differentiable layer is used in the middle of a model (rather than at the start), the model will raise an exception related to differentiability when trying to compute gradients (e.g. as part of `fit`).
+
+
+### New layers
+
+- `PreprocessingLayer` base class: implements shared logic, in particular the `adapt` method for setting the state of the layer.
+- `PreprocessingStage` class: makes it possible to chain multiple preprocessing layers together while training them in one single `adapt` call (by doing cascading training of the underlying layers).
+- `Normalization`: normalizes data feature-wise by subtracting the mean of some sample dataset and dividing by the variance.
+- `Discretization`: transforms continuous data into one-hot encoded binary vectors representing the different "bins" that the continuous data belongs to.
+- `TextVectorization`: transforms string data into either dense vectors (e.g. TF-IDF transform) or sequences of token indices (e.g. to be passed to an `Embedding` layer).
+
+
+## Design details
+
+### Detailed layer signatures
+
+#### PreprocessingLayer
+
+```python
+def adapt(self, data, reset_state=True):
+    """Fits the state of the preprocessing layer to the data being passed.
+
+    Arguments:
+        data: The data to train on. It can be passed either as a tf.data Dataset,
+            or as a numpy array (or a dict or list of arrays in case of multi-input
+            preprocessing stages).
+        reset_state: Optional argument specifying whether to clear the state of the
+            layer at the start of the call to `adapt`, or whether to start from
+            the existing state. This argument may not be relevant to all
+            preprocessing layers: a subclass of PreprocessingLayer may chose to
+            only implement `adapt(self, data)`.
+    """
+```
+
+#### PrepocessingStage
+
+There are two ways to instantiate a `PrepocessingStage` layer: either `Sequential` style (pass a list of preprocessing layer instances) or Functional style (pass the inputs and outputs of a DAG of preprocessing layers).
+
+If any layer other than `PreprocessingLayer` instances is included in a `PrepocessingStage`, these layers will be treated as frozen both during `adapt` and later during `fit`.
+
+
+#### Normalization
+
+```python
+def __init__(self, axis=-1, **kwargs):
```

Comment:
I think users should optionally be able to specify `mean` and `std` parameters to be used for this layer? Default can be adapting to the given dataset, but, it should also be possible for me to say what mean and std to use.

Reply:
I believe this was answered in https://github.com/keras-team/governance/pull/1#issuecomment-490259942 . Are you thinking about another use case?

Reply:
Thanks. Yes answers.